### PR TITLE
Extend reroute with an option to force assign stale primary shard copies

### PR DESCRIPTION
**Pull request by ywelsch**
Relates to #14739.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+        public AllocatePrimaryAllocationCommand fromXContent(XContentParser parser) throws IOException {
+            String index = null;
+            int shardId = -1;
+            String nodeId = null;
+            String allocationId = null;
+            boolean ignoreDeciders = false;
+
+            String currentFieldName = null;
+            XContentParser.Token token;
+            while ((token = parser.nextToken()) != XContentParser.Token.END_OBJECT) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: Should this use the `ObjectParser`? I think everything is going that way.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        }
+
+        @Override
+        public void toXContent(AllocatePrimaryAllocationCommand command, XContentBuilder builder, ToXContent.Params params, String objectName) throws IOException {
+            if (objectName == null) {
+                builder.startObject();
+            } else {
+                builder.startObject(objectName);
+            }
+            builder.field("index", command.shardId().index().name());
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: Other places we use a `Fields` inner class for the field names. Not here? I'm not really 100% sure why we don't just use strings for the field names so I can't speak with confidence about this.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+        DiscoveryNode discoNode = null;
+        if (node != null) {
+            discoNode = allocation.nodes().resolveNode(node);
+        }
+        final RoutingNodes routingNodes = allocation.routingNodes();
+
+        // check if we have a primary unassigned shard
+        ShardRouting shardRouting = null;
+        for (ShardRouting routing : routingNodes.unassigned()) {
+            if (routing.shardId().equals(shardId)) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: Smoosh these into one if statement? Its not a requirement, I'm more curious why you didn't do it.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+                return getFailedRerouteExplanation(explain, allocation,
+                    "Allocation can only be done on data nodes, not [" + node + "]");
+            } else {
+                return getFailedRerouteExplanation(explain, allocation,
+                    "Could not find [" + node + "] among the routing nodes");
+            }
+        }
+
+        Decision decision = allocation.deciders().canAllocate(shardRouting, routingNode, allocation);
+        if (ignoreDeciders == false && decision.type() == Decision.Type.NO) {
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: Why even ask if `ignoreDeciders == true`?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java (hunk 1 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: I essentially copy pasted this stuff from the `AllocateAllocationCommand`. You are right though, I should transform it to use `ObjectParser`.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+        for (RoutingNodes.UnassignedShards.UnassignedIterator it = routingNodes.unassigned().iterator(); it.hasNext(); ) {
+            ShardRouting unassigned = it.next();
+            if (unassigned != shardRouting) {
+                continue;
+            }
+            it.initialize(routingNode.nodeId(), unassigned.version(), allocation.clusterInfo().getShardSize(unassigned, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE));
+            break;
+        }
+        if (ignoreDeciders && decision.type() == Decision.Type.NO) {
+            decision = allocation.decision(Decision.YES, RECOVER_PRIMARY_ALLOCATION_COMMAND, "override deciders");
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: This seems mighty dangerous but I get it. I suspect things will start relocating as soon as you do this too.

Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: It is dangerous, yes. This flag should only be used with extreme care and needs to be documented as such. The reason it's there is to allow manual primary allocation to be as powerful as for the automatic case. In the automatic case, we force allocation even if all nodes say no (see PrimaryShardAllocator: https://github.com/elastic/elasticsearch/blob/c50b22f95f48119213156bdb634925e1be91b794/core/src/main/java/org/elasticsearch/gateway/PrimaryShardAllocator.java#L140 ). As you noted, this can trigger relocation of other shards.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java (hunk 4 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: The idea behind that was so that we could see in the RerouteExplanation if it was Yes, Forced Yes or Throttle. It's probably simpler to just always provide the decision  (YES, "override deciders") in case of explicit override.

**Comment by bleskes** [comment]
@ywelsch and I discussed this yesterday. Here is the gist of the discussion:

1) Instead of fetching the store data in the command, adapt unassigned info so that the node initializing the shard will expect local data and fail if it's not there.
2) Remove the ignore allocation decider param and always assign if the user asked for it (in the case of the primary assignment)
3) Try to fold the new AllocatePrimaryAllocationCommand in the current AllocateAllocationCommand.

Here's a suggest for how the rest layer will look like for that:

```
POST _cluster/reroute
{
  "commands": [
    {
      "allocate": {
        "index": "index",
        "shard": 0,
        "node": "node",
        "allow_primary": "never|existing_only|force_empty"
      }
    }
  ]
```

Note the `allow_primary` option. The default, `never` refuses to allocate a primary. `existing_only` allocates it but must reuse an existing copy. `force_empty` does what true does now - i.e., force the creation of an empty primary. Also the last two should imply ignoring the allocation deciders.

@ywelsch thoughts?

**Comment by ywelsch** [comment]
Before this PR, users only had the possibility to either allocate a replica or an empty primary. The flag "allow_primary" was misleading in that context. I would very much like to get rid of this flag altogether and introduce something new. What about introducing a new "mode" field:

"mode": "replica" // default, only allocates unassigned replica shard (fails if unassigned primary exists)
"mode": "primary" // only allocates if primary unassigned and if data available on node
"mode": "empty_primary" // only allocates if primary unassigned, creates empty shard

"mode": "replica" is equivalent to the old "allowPrimary" : "false"
"mode": "empty_primary" is equivalent to the old "allowPrimary" : "true"

**Comment by ywelsch** [comment]
Also, relates to #14352

**Comment by bleskes** [comment]
I think mode is a bit ambiguous as a flag. I do understand the concerns that `allow_primary` will now do something different than the allow primary in the cancel command.  As far as the values. I like "existing_only" better than just "primary" as it better communicates the restriction (we must use local data). Also I like "force_empty" better than "empty_primary" because it communicates the implications better and the severity better. It also better describes what we do today - always create an empty shard, even if there is local data. That last one is of course subject to change (but that's a bigger discussion)

**Comment by ywelsch** [comment]
>  As far as the values. I like "existing_only" better than just "primary" as it better communicates the restriction (we must use local data). Also I like "force_empty" better than "empty_primary" because it communicates the implications better and the severity better. It also better describes what we do today - always create an empty shard, even if there is local data.

```
$ grep -o "better" | wc -l
6
```

sorry ;-)

**Comment by clintongormley** [comment]
We have a stalled issue (https://github.com/elastic/elasticsearch/issues/14352) about choosing a better name for `allow_primary` which properly indicates the fact that you can lose your data.  The issue is stalled because it is difficult to come up with the right name.

Instead, what about changing the action name from `allocate` to:

* `allocate_replica`
* `allocate_empty_primary` 
* and for this PR: `allocate_stale_primary`

@s1monw is concerned about exposing these last two actions via the API and would prefer to provide a command line tool (perhaps as a plugin) to do this job instead.  The idea is to make it harder for users to make a change that will delete data, and will hopefully force them to read about the consequences of their action.

While I understand this sentiment, I'm not convinced that a command line tool is the way to go here.  ES is API driven so this would be quite a change from what we do today, and there are probably many users who don't have command line access to their hosted clusters, and who need to fix a problem quickly without having to wait for support to respond to their tickets.

Instead, I'd propose adding an `accept_data_loss` flag to these actions which must be set to true for the action to be taken.  I think this would be enough to warn the user that they're doing something dangerous.

**Comment by bleskes** [comment]
+1. My only suggestion would be to use 'force_empty_primary' and 'force_stale_primary' as command names, to make it even clearer that (all) shard content will be lost.

**Comment by ywelsch** [comment]
@bleskes @clintongormley I pushed a new set of changes reflecting the naming suggested by @clintongormley. This new changeset also contains documentation for the new commands. Two open points remain for me:

-  I am still not fully convinced about the command naming. I like @clintongormley's suggestion of command names `allocate_replica`, `allocate_empty_primary` and `allocate_stale_primary`. The `accept_data_loss` flag, however, feels unnatural to me in the rest api and even weirder in the Java API. I am not completely against a safeguard, but would prefer to have it in another form. As an inspiration, we might look at the way it is done for wildcard indices deletion ([`action.destructive_requires_name`](https://www.elastic.co/guide/en/elasticsearch/reference/current/indices-delete-index.html)).
- To keep the commands symmetric, we should also split the `cancel` command (which currently has an `allowPrimary` flag) into two commands `cancel_replica` and `cancel_primary`.

@bleskes What I don't like about `force_empty_primary` is that the action verb `allocate` is lost.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        return builder;
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(shardId.getIndex());
+        out.writeVInt(shardId.getId());
+        out.writeString(node);
+    }
+
+    public static abstract class Factory<T extends AbstractAllocateAllocationCommand> implements AllocationCommand.Factory<T> {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: Since this is a major rewrite anyway, can we try to move from custom factories  to `NamedWriteableRegistry` ?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+     * @return id of the node
+     */
+    public String node() {
+        return this.node;
+    }
+
+    /**
+     * Gets reroute explanation for a routing node that cannot be found
+     */
+    protected RerouteExplanation getMissingRoutingNodeExplanation(RoutingAllocation allocation, boolean explain, DiscoveryNode discoNode) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: can call this `explainOrThrowMissingRoutingNode` ? the docs can read something like "a utility method to handle the case where a disco node can not be found in the routing table. Typically this would mean it's not a data node"

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+            return getNegativeRerouteExplanation(explain, allocation, "Allocation can only be done on data nodes, not [" + node + "]");
+        } else {
+            return getNegativeRerouteExplanation(explain, allocation, "Could not find [" + node + "] among the routing nodes");
+        }
+    }
+
+    /**
+     * Gets the first unassigned primary shard routing for the current shard id. Returns null if it can't find any.
+     */
+    protected ShardRouting getUnassignedPrimaryShardRouting(RoutingNodes routingNodes) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: Callers of this method can just do `allocation.routingTable().shardRoutingTable(shardId).primaryShard()` these days (if we add an overload for shardRoutingTable which takes a shardId). I don't think it's worth having this utility method. (I know it existed before - we have progressed since it was written :))

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+     */
+    protected void initializeUnassignedShard(RoutingAllocation allocation, RoutingNodes routingNodes, RoutingNode routingNode, ShardRouting shardRouting) {
+        for (RoutingNodes.UnassignedShards.UnassignedIterator it = routingNodes.unassigned().iterator(); it.hasNext(); ) {
+            ShardRouting unassigned = it.next();
+            if (unassigned != shardRouting) {
+                continue;
+            }
+            it.initialize(routingNode.nodeId(), unassigned.version(), allocation.clusterInfo().getShardSize(unassigned, ShardRouting.UNAVAILABLE_EXPECTED_SHARD_SIZE));
+            break;
+        }
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: can we assert we found the requested shard in the loop?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+                }
+            }
+        }
+        return null;
+    }
+
+    /**
+     * Creates negative reroute explanation based on provided text
+     */
+    protected RerouteExplanation getNegativeRerouteExplanation(boolean explain, RoutingAllocation allocation, String text) {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: same - please name it something like `explainOrThrowRejectedCommand`

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+        if (explain) {
+            return new RerouteExplanation(this, allocation.decision(Decision.NO, name() + "_allocation_command", text));
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: maybe just use `name() + " (allocation command)"` ?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java (hunk 10 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: another nit: call the `text` param, reason?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java (hunk 6 above)
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: I see now that MoveAllocationCommand is not touched by the PR. I think moving to NamedWriteableRegistry is a good idea, but I'm fine with putting it out of scope for this PR

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+        final DiscoveryNode discoNode = allocation.nodes().resolveNode(node);
+        final RoutingNodes routingNodes = allocation.routingNodes();
+        RoutingNode routingNode = routingNodes.node(discoNode.id());
+        if (routingNode == null) {
+            return getMissingRoutingNodeExplanation(allocation, explain, discoNode);
+        }
+
+        ShardRouting shardRouting = getUnassignedPrimaryShardRouting(routingNodes);
+        if (shardRouting == null) {
+            return getNegativeRerouteExplanation(explain, allocation, "failed to find primary " + shardId + " on the list of unassigned shards");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java]: maybe the reason can simply be "primary " + shardId + " is already assigned".

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java [hunk 13]
```diff
@@ -0,0 +1,112 @@
...
+        }
+
+        if (acceptDataLoss == false) {
+            return getNegativeRerouteExplanation(explain, allocation,
+                "trying to allocate an empty primary shard " + shardId + " that entails possible data loss, needs to be explicitly enabled");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java]: Can this read `allocating an empty primary for " + shardId + " can result in data loss. please confirm by setting the accept_data_loss parameter to true.` ?

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateEmptyPrimaryAllocationCommand.java]: also, I don't think we need to do this if unassignedInfo.getReason == INDEX_CREATED?

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.io.IOException;
+
+/**
+ * Abstract base class for allocating an unassigned primary shard to a node
+ */
+public abstract class AllocatePrimaryAllocationCommand extends AbstractAllocateAllocationCommand {
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocatePrimaryAllocationCommand.java]: can we maybe call this BasePrimaryAllocationCommand? got confused while reading the extends clause somewhere else thinking this is a real command until I saw the abstract keyword here.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+
+        ShardRouting primaryShardRouting = getUnassignedPrimaryShardRouting(routingNodes);
+        if (primaryShardRouting != null) {
+            return getNegativeRerouteExplanation(explain, allocation,
+                "trying to allocate a replica shard " + shardId + ", while corresponding primary shard is still unassigned");
+        }
+
+        ShardRouting shardRouting = getUnassignedReplicaShardRouting(routingNodes);
+        if (shardRouting == null) {
+            return getNegativeRerouteExplanation(explain, allocation, "failed to find " + shardId + " on the list of unassigned shards");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateReplicaAllocationCommand.java]: maybe "all copies of " + shardId +" are already assigned. use the move allocation command instead".

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+        final DiscoveryNode discoNode = allocation.nodes().resolveNode(node);
+        final RoutingNodes routingNodes = allocation.routingNodes();
+        RoutingNode routingNode = routingNodes.node(discoNode.id());
+        if (routingNode == null) {
+            return getMissingRoutingNodeExplanation(allocation, explain, discoNode);
+        }
+
+        ShardRouting shardRouting = getUnassignedPrimaryShardRouting(routingNodes);
+        if (shardRouting == null) {
+            return getNegativeRerouteExplanation(explain, allocation, "failed to find primary " + shardId + " on the list of unassigned shards");
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AllocateStalePrimaryAllocationCommand.java]: see text from other suggestion for empty primary allocation

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 17]
```diff
@@ -41,7 +53,7 @@
 @ESIntegTestCase.SuppressLocalMode
 public class PrimaryAllocationIT extends ESIntegTestCase {
 
-    public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception {
+    private void createStaleReplicaScenario() throws Exception {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: nice!

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 18]
```diff
@@ -103,6 +119,48 @@ public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception
...
+        assertHitCount(client().prepareSearch("test").setSize(0).setQuery(matchAllQuery()).get(), useStaleReplica ? 1l : 0l);
+    }
+
+    public void testForcePrimaryShardIfAllocationDecidersSayNoAfterIndexCreation() throws ExecutionException, InterruptedException {
+        String node = internalCluster().startNode();
+        client().admin().indices().prepareCreate("test").setSettings(Settings.builder()
+            .put("index.routing.allocation.exclude._name", node)
+            .put("index.number_of_shards", 1).put("index.number_of_replicas", 0)).get();
+
+        ClusterHealthResponse health = client().admin().cluster()
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: create does an inline reroute, so you can check directly here that no shard is assigned. No need for timeouts

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java [hunk 19]
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
...
         } catch (IllegalArgumentException e) {
         }
 
-        logger.info("--> allocating with primary flag set to true");
-        rerouteResult = allocation.reroute(clusterState, new AllocationCommands(new AllocateAllocationCommand(new ShardId("test", 0), "node1", true)));
+        logger.info("--> allocating empty primary with acceptDataLoss flag set to false");
+        try {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateEmptyPrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
+        } catch (IllegalArgumentException e) {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: can we check the message here?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java [hunk 20]
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
...
             fail();
         } catch (IllegalArgumentException e) {
         }
 
-        logger.info("--> allocating with primary flag set to true");
-        rerouteResult = allocation.reroute(clusterState, new AllocationCommands(new AllocateAllocationCommand(new ShardId("test", 0), "node1", true)));
+        logger.info("--> allocating empty primary with acceptDataLoss flag set to false");
+        try {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateEmptyPrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: can you add a message to this fail?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java [hunk 21]
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
...
+        } catch (IllegalArgumentException e) {
+        }
+
+        logger.info("--> allocating stale primary with acceptDataLoss flag set to false");
+        try {
+            allocation.reroute(clusterState, new AllocationCommands(new AllocateStalePrimaryAllocationCommand(new ShardId("test", 0), "node1", false)));
+            fail();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: same as above

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java [hunk 22]
```diff
@@ -122,22 +133,36 @@ public void testAllocateCommand() {
...
+        } catch (IllegalArgumentException e) {
+        }
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: same as above

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java [hunk 23]
```diff
@@ -210,7 +235,7 @@ public void testCancelCommand() {
         assertThat(clusterState.getRoutingNodes().shardsWithState(INITIALIZING).size(), equalTo(0));
 
         logger.info("--> allocating with primary flag set to true");
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: stale logging message - use allow data loss set to true

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 24]
```diff
@@ -103,6 +119,48 @@ public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception
...
+            .health(Requests.clusterHealthRequest("test").timeout(TimeValue.timeValueSeconds(1)).waitForGreenStatus()).get();
+        if (health.isTimedOut() == false) {
+            fail("Index should not be allocated due to exclude constraint");
+        }
+
+        client().admin().cluster().prepareReroute().add(new AllocateEmptyPrimaryAllocationCommand(new ShardId("test", 0), node, true)).get();
+        ensureGreen("test");
+    }
+
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: can we add a variant that tests that assigning a stale primary to a node that doesn't have the data fails to initialize the shard?

File: docs/reference/cluster/reroute.asciidoc [hunk 25]
```diff
@@ -64,14 +64,41 @@ The commands supported are:
...
+the cluster. To prevent data loss, the system does not automatically promote a stale
+shard copy to primary.
+
+As a manual override, two commands to forcefully allocate primary shards
+are available:
+
+`allocate_stale_primary`::
+    Allocate an unassigned stale primary shard to a node. Accepts the
+    `index` and `shard` for index name and shard number, and `node` to
+    allocate the shard to. Using this command leads to a loss of data
```
Comment (bleskes) [inline docs/reference/cluster/reroute.asciidoc]: may lead to data loss..

File: docs/reference/cluster/reroute.asciidoc [hunk 26]
```diff
@@ -64,14 +64,41 @@ The commands supported are:
...
+- A new index was created but there is no node which satisfies the allocation deciders.
+- An up-to-date shard copy of the data cannot be found on the current data nodes in
+the cluster. To prevent data loss, the system does not automatically promote a stale
+shard copy to primary.
+
+As a manual override, two commands to forcefully allocate primary shards
+are available:
+
+`allocate_stale_primary`::
+    Allocate an unassigned stale primary shard to a node. Accepts the
```
Comment (bleskes) [inline docs/reference/cluster/reroute.asciidoc]: Allocates a primary shard to a node that holds a stale copy.

File: docs/reference/cluster/reroute.asciidoc [hunk 27]
```diff
@@ -64,14 +64,41 @@ The commands supported are:
...
+    data rejoins the cluster later on, that data will be overwritten with
+    the data of the stale copy that was forcefully allocated with this
+    command. To ensure that these implications are well-understood,
+    this command requires the special field `accept_data_loss` to be
+    explicitly set to `true` for it to work.
+
+`allocate_empty_primary`::
+    Allocate an unassigned empty primary shard to a node. Accepts the
+    `index` and `shard` for index name and shard number, and `node` to
+    allocate the shard to. Using this command leads to a complete loss
```
Comment (bleskes) [inline docs/reference/cluster/reroute.asciidoc]: Using this command leads to a complete loss of all data that was indexed into this shard, if it was previously started.

**Comment by bleskes** [comment]
Thanks @ywelsch . I really like how it turned out. Left some minor comments here and there.

File: core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java (hunk 8 above)
Comment (ywelsch) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: This does not work unfortunately. The issue is that the ShardRouting instance in the RoutingTable is a different instance than the one in RoutingNodes. `initializeUnassignedShard` which uses reference equality does not match the proper shard then.

Comment (bleskes) [inline core/src/main/java/org/elasticsearch/cluster/routing/allocation/command/AbstractAllocateAllocationCommand.java]: hmm. That smells. How about using `ShardRouting#equalsIgnoringMetaData` there?

**Comment by ywelsch** [comment]
pushed another round of changes addressing the comments. Please have another look @bleskes. (The `NamedWriteableRegistry` refactoring will be for another time)

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java [hunk 28]
```diff
@@ -103,6 +116,63 @@ public void testDoNotAllowStaleReplicasToBePromotedToPrimary() throws Exception
         assertHitCount(client().prepareSearch().setSize(0).setQuery(matchAllQuery()).get(), 2l);
     }
 
+    public void testFailedAllocationOfStalePrimaryToDataNodeWithNoData() throws Exception {
+        createStaleReplicaScenario();
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: I think this can be made simpler by just creating a one node cluster, create an index with no replicas, add a node to the cluster. Shutdown the first node and force assign. Makes sense?

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java [hunk 29]
```diff
@@ -107,9 +121,10 @@ public void testAllocateCommand() {
                 .put(IndexMetaData.builder("test").settings(settings(Version.CURRENT)).numberOfShards(1).numberOfReplicas(1))
                 .build();
         RoutingTable routingTable = RoutingTable.builder()
-                .addAsNew(metaData.index("test"))
+                .addAsRecovery(metaData.index("test"))
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: why was this needed?

File: core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java (hunk 28 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/PrimaryAllocationIT.java]: ok

**Comment by bleskes** [comment]
LGTM. Made one suggestion for simplifying a test (and asked a question). Feel free to adopt the suggestion (or not) and push. Thanks for all the iterations, @ywelsch .

File: core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java (hunk 29 above)
Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: because later we test that allocating empty primary with accept_data_loss flag set to false fails.

Comment (bleskes) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: kk. cool. Can you add a comment?

Comment (ywelsch) [inline core/src/test/java/org/elasticsearch/cluster/routing/allocation/AllocationCommandsTests.java]: will do

**Comment by ywelsch** [comment]
Thanks for the thorough review @bleskes !
